tools/sandbox/src/commands/run/profile.rs

```diff
@@ -0,0 +1,167 @@
+//! a symbol name using a symbol map built using the DWARF information in the binary,
+//! which must be compiled in debug mode.
+
+use std::error::Error;
+use std::fs;
+use std::ops::Bound;
+use std::sync::Arc;
+use std::sync::Mutex;
+use std::sync::mpsc;
+use std::thread;
+use std::time::Duration;
+
+use octez_riscv::machine_state::block_cache::DefaultCacheConfig;
+use octez_riscv::machine_state::memory;
+use octez_riscv::pvm::hooks::NoHooks;
+use octez_riscv::state_backend::owned_backend::Owned;
+use octez_riscv::stepper::StepResult;
+use octez_riscv::stepper::Stepper;
+use octez_riscv::stepper::StepperStatus;
+use octez_riscv::stepper::pvm::PvmStepper;
+use tezos_smart_rollup::types::SmartRollupAddress;
+use tezos_smart_rollup::utils::inbox::InboxBuilder;
+
+use super::BlockImpl;
+use super::RunOptions;
+use super::sample::Sampler;
+use super::sample::Symbols;
+
+/// Run profiling with the specified memory configuration
+pub(crate) fn profile_with_memory_config<MC: memory::MemoryConfig>(
+    opts: RunOptions,
+) -> Result<(), Box<dyn Error>>
+where
+    MC::State<Owned>: Send,
+{
+    let sample_interval = Duration::from_micros(
+        opts.sample_interval_us
+            .expect("Sample interval is expected to be set when profiling"),
+    );
+
+    let program = fs::read(&opts.input)?;
+    let symbols = Symbols::new(&program)?;
+    let mut sampler = Sampler::new(symbols)?;
+
+    let mut inbox = InboxBuilder::new();
+    if let Some(inbox_file) = &opts.common.inbox.file {
+        inbox.load_from_file(inbox_file)?;
+    }
+
+    let rollup_address = SmartRollupAddress::from_b58check(opts.common.inbox.address.as_str())?;
+
+    // The stepper for a normal run is initialised with a [`Console`], which is not [`Send`].
+    // When profiling, the stepper is initialised with [`NoHooks`] instead in order to allow
+    // it to be moved into a separate thread.
+    let stepper = PvmStepper::<NoHooks, MC, DefaultCacheConfig, Owned, BlockImpl<MC>>::new(
+        program.as_slice(),
+        inbox.build(),
+        NoHooks,
+        rollup_address.into_hash().as_ref().try_into()?,
+        opts.common.inbox.origination_level,
+        opts.common.preimage.preimages_dir.clone(),
+        Default::default(),
+    )?;
+
+    let step_limit = opts
+        .common
+        .max_steps
+        .map_or(Bound::Unbounded, Bound::Included);
+    let total_steps = run_stepper_concurrent(stepper, &mut sampler, sample_interval, step_limit)?;
+
+    let sample_count = sampler.finish(&opts.output)?;
+    eprintln!("Total steps: {total_steps}");
+    eprintln!("Samples collected: {sample_count}");
+
+    Ok(())
+}
+
+/// Run the profiling stepper loop with time-based concurrent sampling
+fn run_stepper_concurrent<S>(
+    mut stepper: S,
+    sampler: &mut Sampler,
+    sample_interval: Duration,
+    max_steps: Bound<usize>,
+) -> Result<usize, String>
+where
+    S: Stepper<Manager = Owned> + Send + 'static,
```

Comment:
is the `Send` needed? IIUC, you could equally spin the sampler into the background thread.

Reply:
Right! This also simplifies the running logic, since the same stepper can be run either normally or for profiling. Done [here](https://github.com/tezos/riscv-pvm/compare/86572e492d725416e7f1235e79593c3904f1897e..2101a9f127e2c4495f18487f32608908eb1b36b2).